File: solutions/typescript/problems/[0061_Medium] Rotate List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function rotateRight(head: ListNode | null, k: number): ListNode | null {
    if (head === null) return null;

    // Make the linked-list circular
    let tail: ListNode | null = head;
    let count = 1;
    while (tail.next) {
        tail = tail.next;
        count++;
    }
    tail.next = head;

    // Define the cut
    let curr: ListNode = head;
    let next: ListNode = head.next!;
    let rotations = count - 1 - (k % count);
    while (rotations > 0) {
        curr = next;
        next = next.next!;
        rotations--;
    }
    curr!.next = null;
    return next;
}

File: solutions/c++/problems/[0061_Medium] Rotate List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *rotateRight(ListNode *head, int k) {
        if (!head) return nullptr;
        int size = 0;
        auto curr = head;
        ListNode *last;
        while (curr) {
            if (!curr->next) last = curr;
            size++;
            curr = curr->next;
        }
        last->next = head;
        k %= size;
        int count = size - k - 1;
        curr = head;
        while (count > 0) {
            count--;
            curr = curr->next;
        }
        auto newHead = curr->next;
        curr->next = nullptr;
        return newHead;
    }
};